File: ureka-framework-main/ureka_framework/logic/stage_worker/measure_helper.py
# Environment
from ureka_framework.environment import Environment

# Data Model (RAM)
from ureka_framework.model.shared_data import SharedData

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Resource (Measurer)
from ureka_framework.resource.logger.simple_measurer import measure_worker_func

# Resource (Measurer)
from ureka_framework.resource.logger.simple_measurer import (
    start_process_timer,
    start_perf_timer,
    get_process_time,
    get_perf_time,
    start_comm_timer,
    get_comm_time,
    simple_size_calculator,
)


class MeasureHelper:
    def __init__(self, shared_data: SharedData) -> None:
        self.shared_data = shared_data

    ######################################################
    # Measurement Helper:
    #   Process Response Time
    ######################################################
    @measure_worker_func
    def measure_process_perf_start(self) -> None:
        start_process_timer()
        start_perf_timer()

    @measure_worker_func
    def measure_recv_cli_perf_time(self, cli_name: str) -> None:
        # Response Time
        cli_process_time: float = get_process_time()
        cli_perf_time: float = get_perf_time()
        cli_blocked_time: float = abs(cli_perf_time - cli_process_time)

        # Collect Measurement Raw Data
        if self.shared_data.measure_rec.get(cli_name) is None:
            self.shared_data.measure_rec[cli_name] = {}
        # Record cli_perf_time
        self.shared_data.measure_rec[cli_name]["cli_perf_time"] = cli_perf_time
        self.shared_data.measure_rec[cli_name]["cli_blocked_time"] = cli_blocked_time
        # simple_log("measure", f"measure_rec = {self.shared_data.measure_rec}")

        # Print Measurement Raw Data
        simple_log("measure", f"")
        simple_log("measure", f"+ Receive CLI Input: {cli_name}")
        simple_log(
            "measure",
            f"cli_perf_time = {cli_perf_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
        )
        if cli_blocked_time > Environment.IO_BLOCKING_TOLERANCE_TIME:
            simple_log("warning", f"+ PROC I/O MAYBE BLOCKED TOO LONG...")
            simple_log(
                "warning",
                f"cli_blocked_time = {cli_blocked_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
            )

    @measure_worker_func
    def measure_recv_msg_perf_time(self, comm_name: str) -> None:
        # Response Time
        msg_process_time: float = get_process_time()
        msg_perf_time: float = get_perf_time()
        msg_blocked_time: float = abs(msg_perf_time - msg_process_time)

        # Collect Measurement Raw Data
        if self.shared_data.measure_rec.get(comm_name) is None:
            self.shared_data.measure_rec[comm_name] = {}
        # Record msg_perf_time
        self.shared_data.measure_rec[comm_name]["msg_perf_time"] = msg_perf_time
        self.shared_data.measure_rec[comm_name]["msg_blocked_time"] = msg_blocked_time
        # Record cached comm_time & message_size
        if (
            self.shared_data.measure_rec.get("_recv_message") is not None
            and self.shared_data.measure_rec.get("_recv_message").get("comm_time")
            is not None
        ):
            self.shared_data.measure_rec[comm_name][
                "comm_time"
            ] = self.shared_data.measure_rec["_recv_message"]["comm_time"]
        if (
            self.shared_data.measure_rec.get("_recv_message") is not None
            and self.shared_data.measure_rec.get("_recv_message").get("message_size")
            is not None
        ):
            self.shared_data.measure_rec[comm_name][
                "message_size"
            ] = self.shared_data.measure_rec["_recv_message"]["message_size"]
        # simple_log("measure", f"measure_rec = {self.shared_data.measure_rec}")

        # Print Measurement Raw Data
        simple_log(
            "measure",
            f"msg_perf_time = {msg_perf_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
        )
        if msg_blocked_time > Environment.IO_BLOCKING_TOLERANCE_TIME:
            simple_log("warning", f"+ PROC I/O MAYBE BLOCKED TOO LONG...")
            simple_log(
                "warning",
                f"msg_blocked_time = {msg_blocked_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
            )
        simple_log("measure", f"+ Receive Message Input: {comm_name}")
        simple_log("measure", f"")

    ######################################################
    # Measurement Helper:
    #   Comm Response Time
    ######################################################
    @measure_worker_func
    def measure_comm_perf_start(self) -> None:
        start_comm_timer()

    @measure_worker_func
    def measure_comm_time(self, comm_name: str) -> None:
        # Response Time
        comm_time: float = get_comm_time()

        # Collect Measurement Raw Data
        if self.shared_data.measure_rec.get(comm_name) is None:
            self.shared_data.measure_rec[comm_name] = {}
        # Cache comm_time
        self.shared_data.measure_rec[comm_name]["comm_time"] = comm_time
        # simple_log("measure", f"measure_rec = {self.shared_data.measure_rec}")

        # Print Measurement Raw Data
        simple_log("measure", f"")
        simple_log("measure", f"+ Receive Comm Input: {comm_name}")
        simple_log(
            "measure",
            f"comm_time = {comm_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
        )
        if comm_time > Environment.COMM_BLOCKING_TOLERANCE_TIME:
            simple_log("warning", f"+ COMM I/O MAYBE BLOCKED TOO LONG...")
            simple_log(
                "warning",
                f"comm_time = {comm_time:{Environment.MEASUREMENT_TIME_PRECISION}} seconds",
            )

    ######################################################
    # Measurement Helper:
    #   Data Size
    ######################################################
    @measure_worker_func
    def measure_message_size(
        self, comm_name: str, received_message_with_header: str
    ) -> None:
        # Data Size
        message_size: int = simple_size_calculator(received_message_with_header)

        # Collect Measurement Raw Data
        if self.shared_data.measure_rec.get(comm_name) is None:
            self.shared_data.measure_rec[comm_name] = {}
        # Cache message_size
        self.shared_data.measure_rec[comm_name]["message_size"] = message_size
        # simple_log("measure", f"measure_rec = {self.shared_data.measure_rec}")

        # Print Measurement Raw Data
        simple_log("measure", f"message_size = {message_size} bytes")

File: ureka-framework-java/app/src/main/java/ureka/framework/logic/pipeline_flow/stage_worker/MeasureHelper.java
package ureka.framework.logic.pipeline_flow.stage_worker;

import java.util.HashMap;

import ureka.framework.Environment;
import ureka.framework.model.SharedData;
import ureka.framework.resource.logger.SimpleLogger;
import ureka.framework.resource.logger.SimpleMeasurer;

public class MeasureHelper {
    private SharedData sharedData;

    public MeasureHelper(SharedData sharedData) {
        this.sharedData = sharedData;
    }

    public SharedData getSharedData() {
        return sharedData;
    }

    public void setSharedData(SharedData sharedData) {
        this.sharedData = sharedData;
    }

    //////////////////////////////////////////////////////
    // Measurement Helper:
    // Process Response Time
    //////////////////////////////////////////////////////
    public void measureProcessPerfStart() {
        SimpleMeasurer.measureWorkerFunc(this::_measureProcessPerfStart);
    }
    private void _measureProcessPerfStart() {
        SimpleMeasurer.startProcessTimer();
        SimpleMeasurer.startPerfTimer();
    }

    public void measureRecvCliPerfTime(String cliName) {
        SimpleMeasurer.measureWorkerFunc(this::_measureRecvCliPerfTime, cliName);
    }
    private void _measureRecvCliPerfTime(String cliName) {
        // Response Time
        long cliProcessTime = SimpleMeasurer.getProcessTime();
        long cliPerfTime = SimpleMeasurer.getPerfTime();
        long cliBlockedTime = Math.abs(cliPerfTime - cliProcessTime);

        // Collect Measurement Raw Data
        this.sharedData.getMeasureRec().computeIfAbsent(cliName, k -> new HashMap<>());

        // Record cliPerfTime
        this.sharedData.getMeasureRec().get(cliName).put("cliPerfTime", cliPerfTime);
        this.sharedData.getMeasureRec().get(cliName).put("cliBlockedTime", cliBlockedTime);
        // simpleLog("measure", f"measureRec = {self.sharedData.measureRec}")

        // Print Measurement Raw Data
        SimpleLogger.simpleLog("measure", "");
        SimpleLogger.simpleLog("measure", "+ Receive CLI Input: {" + cliName + "}");
        SimpleLogger.simpleLog("measure", "cliPerfTime = " + cliPerfTime + " nanoseconds");
        // "cliPerfTime = {cliPerfTime:{Environment.MEASUREMENT_TIME_PRECISION}} seconds"
        if (cliBlockedTime > Environment.IO_BLOCKING_TOLERANCE_TIME * 1e6 /*Conversion from millisecond to nanosecond*/) {
            SimpleLogger.simpleLog("warning", "+ PROC I/O MAYBE BLOCKED TOO LONG...");
            SimpleLogger.simpleLog(
                "warning",
                "cliPerfTime = " + cliBlockedTime + " nanoseconds"
            );
        }
    }

    public void measureRecvMsgPerfTime(String commName) {
        SimpleMeasurer.measureWorkerFunc(this::_measureRecvMsgPerfTime, commName);
    }
    private void _measureRecvMsgPerfTime(String commName) {
        // Response Time
        long msgProcessTime = SimpleMeasurer.getProcessTime();
        long msgPerfTime = SimpleMeasurer.getPerfTime();
        long msgBlockedTime = Math.abs(msgPerfTime - msgProcessTime);

        // Collect Measurement Raw Data
        this.sharedData.getMeasureRec().computeIfAbsent(commName, k -> new HashMap<>());

        // Record cliPerfTime
        this.sharedData.getMeasureRec().get(commName).put("msgPerfTime", msgPerfTime);
        this.sharedData.getMeasureRec().get(commName).put("msgBlockedTime", msgBlockedTime);

        // Record cached commTime & messageSize
        if (this.sharedData.getMeasureRec().get("_recvMessage") != null
            && this.sharedData.getMeasureRec().get("_recvMessage").get("commTime") != null) {
            this.sharedData.getMeasureRec().get(commName)
                .put("commTime", this.sharedData.getMeasureRec().get("_recvMessage").get("commTime"));
        }
        if (this.sharedData.getMeasureRec().get("_recvMessage") != null
            && this.sharedData.getMeasureRec().get("_recvMessage").get("messageSize") != null) {
            this.sharedData.getMeasureRec().get(commName)
                .put("messageSize", this.sharedData.getMeasureRec().get("_recvMessage").get("messageSize"));
        }
        // simpleLog("measure", f"measureRec = {self.sharedData.measureRec}")

        // Print Measurement Raw Data
        SimpleLogger.simpleLog("measure", "msgPerfTime = " + msgPerfTime + " nanoseconds");
        // "cliPerfTime = {cliPerfTime:{Environment.MEASUREMENT_TIME_PRECISION}} seconds"
        if (msgBlockedTime > Environment.IO_BLOCKING_TOLERANCE_TIME * 1e6 /*Conversion from millisecond to nanosecond*/) {
            SimpleLogger.simpleLog("warning", "+ PROC I/O MAYBE BLOCKED TOO LONG...");
            SimpleLogger.simpleLog("warning", "msgPerfTime = " + msgPerfTime + " nanoseconds");
        }
        SimpleLogger.simpleLog("measure", "+ Receive Message Input = " + commName);
        SimpleLogger.simpleLog("measure", "");
    }
    //////////////////////////////////////////////////////
    // Measurement Helper:
    // Comm Response Time
    //////////////////////////////////////////////////////
    public void measureCommPerfStart() {
        SimpleMeasurer.measureWorkerFunc(this::_measureCommPerfStart);
    }
    private void _measureCommPerfStart() {
        SimpleMeasurer.startCommTimer();
    }

    public void measureCommTime(String commName) {
        SimpleMeasurer.measureWorkerFunc(this::_measureCommTime, commName);
    }
    private void _measureCommTime(String commName) {
        // Response Time
        long commTime = SimpleMeasurer.getCommTime();

        // Collect Measurement Raw Data
        this.sharedData.getMeasureRec().computeIfAbsent(commName, k -> new HashMap<>());
        // Cache commTime
        this.sharedData.getMeasureRec().get(commName).put("commTime", commTime);
        // simpleLog("measure", f"measureRec = {self.sharedData.measureRec}")

        // Print Measurement Raw Data
        SimpleLogger.simpleLog("measure", "");
        SimpleLogger.simpleLog("measure", "+ Receive Comm Input: " + commName);
        SimpleLogger.simpleLog("measure", "commTime = " + commTime + " nanoseconds");
        if (commTime > Environment.COMM_BLOCKING_TOLERANCE_TIME * 1e6/*Conversion from millisecond to nanosecond*/) {
            SimpleLogger.simpleLog("warning", "+ COMM I/O MAYBE BLOCKED TOO LONG...");
            SimpleLogger.simpleLog("warning", "commTime = " + commTime + " nanoseconds");
        }
    }

    //////////////////////////////////////////////////////
    // Measurement Helper:
    // Data Size
    //////////////////////////////////////////////////////
    public void measureMessageSize(String commName, String receivedMessageWithHeader) {
        SimpleMeasurer.measureWorkerFunc(this::_measureMessageSize, commName, receivedMessageWithHeader);
    }
    private void _measureMessageSize(String commName, String receivedMessageWithHeader) {
        // Data Size
        int messageSize = SimpleMeasurer.simpleSizeCalculator(receivedMessageWithHeader);

        // Collect Measurement Raw Data
        this.sharedData.getMeasureRec().computeIfAbsent(commName, k -> new HashMap<>());
        // Cache messageSize
        this.sharedData.getMeasureRec().get(commName).put("messageSize", (long) messageSize);
        // simpleLog("measure", f"measureRec = {self.sharedData.measureRec}")

        // Print Measurement Raw Data
        SimpleLogger.simpleLog("measure", "messageSize = " + messageSize + " bytes");
    }
}
